### ISSUE TITLE
Change "free outputs" to also return `MemoryDataSet` entries from the catalog

### ISSUE THREAD

**Issue by merelcht** [kedro-org/kedro#1900]
## Description
`session.run()` currently returns "free outputs". What `free_output` means in the code is just "output that's not defined in the catalog", which is a subset of "output that's not a MemoryDataSet".

https://github.com/kedro-org/kedro/blob/f4914201d7f6f38318c2c1f074fcdf802b3e1e0d/kedro/runner/runner.py#L78-L91

There was agreement that the "free outputs" output from session isn't very clear. It was suggested to simply return all output from nodes that is not consumed, even if it's defined in the catalog. However, this could lead to very large amounts of data being returned. Instead we'll change it to return all free outputs and additionally any MemoryDataSets that are defined in the catalog.

## Context
https://github.com/kedro-org/kedro/issues/1802

**Comment by datajoely** [comment]
The last line - are those explicitly defined MemoryDataSets or implicit ones?

**Comment by noklam** [comment]
I think this is a bug fix, rather than any behavioral changes. This happens if someone put MemoryDataSet in the catalog and `session.run` will remove it from the output,

```python
free_outputs = pipeline.outputs() - set(catalog.list()) 
# This will be changed to
free_outputs = pipelines.outputs() - set(catalog_excluding_memory_dataset.list())
```

**Comment by noklam** [comment]
Already implemented in #3475.
